pkg/metric_aggregation/aggregation.go

```diff
@@ -0,0 +1,224 @@
+	// during WriteAll().
+	// This means that aggregations are recalculated at most once per
+	// HTTP request, but only if the underlying metric values changed
+	// in the meantime
+	cache map[string]metric.Family
+}
+
+// Generate a header for an aggregated metric family
+// This is based on the original metric header but includes
+// the aggregation labels in the help text
+func generateAggregateHeader(name string, metricType metric.Type, help string, aggregationName string) string {
+	header := strings.Builder{}
+	header.WriteString("# HELP ")
+	header.WriteString(name)
+	header.WriteByte(' ')
+	header.WriteString(strings.TrimSuffix(help, "."))
+	header.WriteString(", aggregated by ")
+	header.WriteString(aggregationName)
+	header.WriteString(".\n# TYPE ")
+	header.WriteString(name)
+	header.WriteByte(' ')
+	header.WriteString(string(metricType))
+
+	return header.String()
+}
+
+// Generate a suffix for the aggregated metric family
+// to be appended to the original family name, e.g.
+// aggregating kube_foo by an aggregation named bar should
+// yield a family called kube_foo_sum_by_bar
+func AggregatedMetricNameSuffix(aggregationName string) string {
+	return "_sum_by_" + aggregationName
+}
+
+// Create a new aggregated metric
+func NewAggregatedMetric(familyName string, familyType metric.Type, familyHelp string, aggregationName string, aggregation Aggregation) AggregatedMetric {
+	aggregatedMetricName := familyName + AggregatedMetricNameSuffix(aggregationName)
+
+	return AggregatedMetric{
+		Header:          generateAggregateHeader(aggregatedMetricName, familyType, familyHelp, aggregationName),
+		Name:            aggregatedMetricName,
+		MetricType:      familyType,
+		LabelNames:      aggregation.LabelNames,
+		Objects:         make(map[types.UID][]AggregatedValue),
+		AggregationFunc: aggregation.Aggregate,
+	}
+}
+
+// Add a new object to the AggregatedMetric
+// We immediately call the aggregation function and store its results
+func (a *AggregatedMetric) Add(obj metav1.Object, family metric.FamilyInterface) {
+	var objects []AggregatedValue
+	family.Inspect(func(f metric.Family) {
+		objects = append(objects, a.AggregationFunc(obj, f)...)
+	})
+	a.Objects[obj.GetUID()] = objects
+}
+
+// Delete an object (by uid) from the set of values to be aggregated
+func (a *AggregatedMetric) Delete(uid types.UID) {
+	delete(a.Objects, uid)
+}
+
+// Delete all objects from the aggregation
+func (a *AggregatedMetric) Clear() {
+	a.Objects = make(map[types.UID][]AggregatedValue)
+}
+
+// Get the result of the aggregation as a metric family
+func (a *AggregatedMetric) GetMetricFamily(shard int) metric.Family {
+	aggValue := make(map[string]float64)
+	labelValueSets := make(map[string][]string)
+	for _, obj := range a.Objects {
+		for _, m := range obj {
+			labelString := fmt.Sprintf("%#v", m.LabelValues)
+			labelValueSets[labelString] = m.LabelValues
+
+			// this is what makes all the aggregations sums
+			aggValue[labelString] += m.Value
```

Comment:
So at the moment, summing is the only aggregation, right? So not vert expensive given that al the floats are in memory.

Reply:
Yes. Memory is the bigger cost (an estimate is ~3x the size of the whole `/metrics` response in extra memory usage). It actually might be better _not_ to cache the aggregation result and add up the numbers every time (trading CPU time and a bit of latency for reduced memory usage).